### PR TITLE
[Tooling] Replace old L10n scripts with new toolkit actions

### PR DESCRIPTION
**Pull request by AliSoftware**
🗂  This is part of project paaHJt-2Ib-p2 

## Why ?

We're in the process of migrating our legacy scripts and tools handling our translations and the related pipeline (extracting strings from code, sending them to GlotPress, downloading translations once ready…) into proper fastlane actions implemented in our release-toolkit.

Now that the latest version of the release-toolkit starts to have most of those new needed actions implemented, this PR is starting to migrate our `Fastfile` to use them instead of the legacy scripts

## Diff breakdown / PR Review Guidance

Might be worth looking at the commits and their message to understand why some files have been removed. TL;DR:

 - Main change is obviously in the `Fastfile` and `Jetpack-Fastfile` files.
   - I've also extracted some of the constants in `lanes/localization-lanes.rb`, in preparation to later extract more things (especially L10n-related lanes) into that file, as explained in item (2) of the "Next Steps" collapsible section down below.
 - `Scripts/extract-framework-translations.swift`, `Scripts/fix-translation` and `Scripts/update-translation.rb` have been removed as they are obsolete now that new release-toolkit actions replace them
 - `InfoPlistTranslator.{h,m}` file removed as it was a half-hack to get the `InfoPlist.strings` strings go into GlotPress — though the translations were never reinjected back on export.
   - This is fully replaced by the `ios_merge_strings_files`/`ios_extract_keys_from_strings_files` strategy (see review comment below)
   - I've moved the translator hints/comments that were provided on those keys from that obsolete `InfoPlistTranslator.m` file to the `en.lproj/InfoPlist.strings` files that are now the new source of truth for those keys (that will be merged with the main one to be imported by GlotPress).
 - A couple of empty and unused `Localizable.strings` files from side targets have been deleted to clean things up (which also allowed me to ensure I didn't forget to merge any other manually-maintained `.strings` file from other targets)
   - `en.lproj/WordPressApi.strings` was not used anywhere and not even part of the `.pbxproj`
   - `WordPressDraftActionExtension`, `WordPressNotificationContentExtension` and `WordPressNotificationServiceExtension` all had empty and unused `Localizable.strings` files that I deleted from the repo and the `.pbxproj`.

I've kept the rest of the changes we discussed in the previous reviews of that PR (implementation of `download_metadata.swift` replacement, reorg the `Fastfile`/`Jetpack-Fastfile` devide, use `fastlane/metadata/default` as source of truth…) for a separate step, so expect other PRs to pop soon on top of this (see "Next Steps" below).

## Test Steps

💡  _Optional: You might want to temporarily switch to the root commit of this PR, run `fastlane generate_strings_file_for_glotpress`, and store aside the `WordPress/Resources/en.lproj/Localizable.strings` file (that gets extracted by `genstrings` from the current state of the codebase), in order to have a point of comparison for how that file looked like when using old actions before that PR, vs the new file that will be generated by the changes in this PR._

 -  ⎇  Cut a new branch from this one; ensure that you're up-to-date with your pods (`rake dependencies` etc)
 -  💻 Run `bundle exec fastlane generate_strings_file_for_glotpress`
 - 🕵️ Verify that the `WordPress/Resources/en.lproj/Localizable.strings` file is updated:
   - Not only with new strings from code since last code freeze,
   - But also contains merged strings from with the app's `InfoPlist.strings`, DraftActionExtension's `InfoPlist.strings` and `WordPressIntents`'s `Sites.strings`, each having prefixes added to their respective keys.
   - And that keys that were previously coming from `InfoPlistTranslator.m` in the `Localizable.strings` (but were never used/extracted with the prior tooling) have been removed, now that the `infoplist.*` keys replace them
   - Also verify that the changes have been commited (and your working copy is not "dirty" and has uncommited left)
 - 🪛  Comment out [lines 476-480 of the `Fastfile`](https://github.com/wordpress-mobile/WordPress-iOS/pull/18061/files#diff-dff4b99d4e651ab9d7597876ec8dbe92aa934e42c0fb55f4aadd6c106fc96688R476-R480), aka the lines calling `ios_download_strings_files_from_glotpress` in the `download_localized_strings_and_metadata` lane. (Optionally also comment the 2 calls to `download_*_localized_app_store_metadata` at the end of the same lane to save you some unnecessary wait during testing)
 - 🪛  Manually edit some random `WordPress/Resources/*.lproj/Localizable.strings` files from a couple of locales other than `en.lproj` — to simulate new translations that would have been coming from the GlotPress download
   - In particular, add dummy translations for the prefixed keys that come from manually-maintained `.strings` files, like `infoplist.NSCameraUsageDescription`, `ios-sharesheet.CFBundleDisplayName` and `ios-widget.BOl9KQ`
 - 💻  Run `bundle exec fastlane download_localized_strings_and_metadata`
 - 🕵️  Ensure that the `WordPress/Resources/*.lproj/InfoPlist.strings`, `WordPress/WordPressDraftActionExtension/*.lproj/InfoPlist.strings` and `WordPress/WordPressIntents/*.lproj/Sites.strings` files — which will now be generated as XML files — have their keys updated accordingly for the locale you added your dummy keys above, with the dummy translations you added manually from previous step, and with the prefixes removed from their keys.
 - 🕵️  Note that the same files for any other locale you didn't add dummy keys to, and which thus won't have the corresponding keys in their `*.lproj/Localizable.strings` file to be extracted, it will result in empty XML plist files, replacing the ones that were present in the repo so far (and in OpenStep format) and had past translation.
   - This disappearance of those past translations is expected, since those strings were not in GlotPress in the first place — they were very probably translated as a one-time event when the corresponding targets were introduced, as part of a one-time Deliverable or similar)
   - This means those strings will have to be re-translated by translators in GlotPress during next iteration. This is expected (and likely not a big deal, and maybe translation memory will also be able to help translators then), so those disappearances of the past translations should not be considered a bug when testing that PR.
   - Also verify that the changes have been commited (and your working copy is not "dirty" and has uncommited left)
 - 🗑️  Delete your test branch

## Next Steps (in future PRs)

<details><summary><h3>1. Finish integration once <tt>gp_downloadmetadata</tt>'s successor is ready</h3></summary>

 - [ ] Replace `download_wordpress_localized_app_store_metadata` and `download_jetpack_localized_app_store_metadata` lanes with a call to the successor of `gp_downloadmetadata`
 - [ ] Remove `fastlane/download_metadata.swift` old script from the repository

</details>

<details><summary><h3>2. Refactor the <tt>Fastfile</tt> vs <tt>Jetpack-Fastfile</tt> split</h3></summary>

See also https://github.com/wordpress-mobile/WordPress-iOS/pull/18061#discussion_r817080826 and https://github.com/wordpress-mobile/WordPress-iOS/pull/18061#issuecomment-1056278845

Especially, doing the separation via the WordPress vs Jetpack dichotomy is probably not the most relevant split anymore, as it actually makes browsing the otherwise related lanes (e.g. `download_{wordpress,jetpack}_localized_app_store_metadata`) more difficult in practice

Instead, it probably makes more sense to move the lanes around to do the split differently, especially focused on the themes of the lanes. Some themes ideas for the new split:
  - `Fastfile-ReleaseLanes` (`code_freeze`, `new_beta_release`, `finalize_release`, …)
  - `Fastfile-Localization` (`generate_strings_file_for_glotpress`, `download_localized_strings_and_metadata`, …)
  - `Fastfile-ASC` (`register_new_device`, `update_certs_and_profiles`, …)
  - `Fastfile-Screenshots` (just rename the existing `ScreenshotsFastfile` to match the convention I'm proposing here)
  - `Fastfile-Helpers` (any helper methods, maybe also some constants)
  - Then the main `Fastfile` would simply `fastlane_required` those, maybe as well as declaring some common constants and `before_all` logic, but that would be it. Making it mostly a facade to all the smaller ones.
</details>

<details><summary><h3>3. Update the Source of truth for the <tt>release-notes.txt</tt> file</h3></summary>

We should make the `fastlane/{metadata,jetpack_metadata}/default/release-notes.txt` be the source of truth where release notes are extracted (from `RELEASE-NOTES.txt`) — instead of using the intermediate `WordPress/Resources/release_notes.txt` and then have to copy them over.

See also discussion in https://github.com/wordpress-mobile/WordPress-iOS/pull/18061#discussion_r817352930

That change will require a separate PR as it will touch different lanes, but also the Release Scenarios, and also the MC Tool we use to generate the announcement post during code-freeze (which looks at those `release_notes.txt` files to include the draft notes in the post)… and maybe more?
</details>

### CODE REVIEW

File: fastlane/Fastfile [hunk 1]
```diff
@@ -458,15 +469,61 @@ platform :ios do
...
-    # translations for the app and App Store Connect metadata.
-    ios_update_metadata(options)
-    # Jetpack: use the custom setup. Notice that we don't need extra steps to
-    # download `.strings` because Jetpack uses the same codebase as WordPress.
+    # Download `Localizable.strings` translations used within the app
+    parent_dir_for_lprojs = File.join(PROJECT_ROOT_FOLDER, 'WordPress', 'Resources')
+    ios_download_strings_files_from_glotpress(
+      project_url: 'https://translate.wordpress.org/projects/apps/ios/dev/', # TODO: Use a constant to DRY stuff
+      # TODO: Fill this in
+      locales: { gp1: lproj1, gp2: lproj2, … }, # TODO: Provide a big constant for this list somewhere. (See `ALL_LANGS` from `update-translation.rb`)
```
Comment (AliSoftware) [inline fastlane/Fastfile]: 🚧  WIP. I'll need to address this before undrafting the PR.

File: fastlane/Fastfile [hunk 2]
```diff
@@ -458,15 +469,61 @@ platform :ios do
...
-    # WordPress: leverage the default release toolkit setup.
-    # It's not clear in the action name, but this downloads both `.strings`
-    # translations for the app and App Store Connect metadata.
-    ios_update_metadata(options)
-    # Jetpack: use the custom setup. Notice that we don't need extra steps to
-    # download `.strings` because Jetpack uses the same codebase as WordPress.
+    # Download `Localizable.strings` translations used within the app
+    parent_dir_for_lprojs = File.join(PROJECT_ROOT_FOLDER, 'WordPress', 'Resources')
+    ios_download_strings_files_from_glotpress(
+      project_url: 'https://translate.wordpress.org/projects/apps/ios/dev/', # TODO: Use a constant to DRY stuff
```
Comment (AliSoftware) [inline fastlane/Fastfile]: This is also used in the lane that checks translation coverage, so there's some DRY-ing to be done here

File: fastlane/Fastfile [hunk 3]
```diff
@@ -458,15 +469,61 @@ platform :ios do
...
+    ios_extract_keys_from_strings_files(
+      source_parent_dir: parent_dir_for_lprojs
+      target_original_files: [
+        File.join(parent_dir_for_lprojs, 'en.lproj', 'InfoPlist.strings'),
+        File.join(parent_dir_for_intents_target_lprojs, 'en.lproj', 'Sites.strings'),
+      ]
+    )
+
+    # Finally, also download the AppStore metadata (app title, keywords, etc.)
+    # @FIXME: Replace this whole lane with a call to the future replacement of `gp_downloadmetadata` once it's implemented in the release-toolkit (see paaHJt-31O-p2).
```
Comment (AliSoftware) [inline fastlane/Fastfile:489]: This FIXME might be for another PR (once `gp_downloadmetadata`'s successor is ready)

Comment (AliSoftware) [reply fastlane/Fastfile:489]: _Braindump on `default` vs `en-US`_

Note that while we're at it (and/or also while we'll be working on updating the keywords soon, see pdjUbD-8v-p2), we might want to take the occasion to revisit the distinction of what goes in `en-US` vs `default` folders.

See also [the docs for `deliver`](https://docs.fastlane.tools/actions/deliver/#available-metadata-folder-options) about this. Typically:
 - `en-US` is the source locale, that contains what we want to send to GlotPress for translation to other locales
 - `default` is used by fastlane for when a given locale folder does not have a given `.txt` file for a field

Some thoughts on our options:
 - Making `default` our source of truth (in which we'd edit the `release_notes.txt` and `keywords.txt` files to put the originals in) could be a nice idea.
 - On the other hand, we might not absolutely want _every_ metadata field to be sent to GlotPress for translation (e.g. `support_url ` or `marketing_url`), so having those not-to-be-translated values in `default`, while having the originals we want to send for translation in `en-US` could make more sense and be more flexible
 - Even though sending even `support_url` and the likes to GlotPress… might not be that big of a deal after all: the translations would currently just be the same as the original, but that would allow us in the future to provide URLs to localized content (e.g. `https://wordpress.com/fr/support`…). After all, those fields are localizable in ASC (which is why `deliver` consider them localizable too), so…

That might also give us the opportunity to use `default` vs `en-US` as a way to differentiate which `.txt` files should go into the `.po` and to GlotPress for translation (`en-US`) and which `.txt` files we don't want to translate (e.g. `keywords.txt` could be moved to `default` during an ASO experiment to stop being included into the `.po`, and moved back into `en-US` once we're ready to translate them again…)

File: fastlane/Fastfile [hunk 4]
```diff
@@ -458,15 +469,61 @@ platform :ios do
...
+  #####################################################################################
+  desc 'Downloads localized WordPress metadata (for App Store Connect) from GlotPress'
+  lane :download_wordpress_localized_app_store_metadata do
+    # @FIXME: Replace this whole lane with a call to the future replacement of `gp_downloadmetadata` once it's implemented in the release-toolkit (see paaHJt-31O-p2).
+
+    # No need to `cd` into `fastlane` because of how Fastlane manages its paths internally.
+    sh './download_metadata.swift wordpress'
+
+    # @TODO: Make the `fastlane/metadata/en-US/release_notes.txt` path be the source of truth for the original copies in the future.
+    # (will require changes in the `update_appstore_strings` lane, the Release Scenario, the MC tool to generate the announcement post…)
```
Comment (AliSoftware) [inline fastlane/Fastfile:510]: Can't wait to do this to simplify the whole thing greatly, as discussed in P2.

But that might require some parallel PRs to make sure such a change is reflected in all the places that depends on that path, including outside the repo and fastlane (Scenario, MC…), so we'll probably keep that for a later step.

File: fastlane/Jetpack-Fastfile [hunk 5]
```diff
@@ -156,41 +156,33 @@ lane :update_jetpack_appstore_strings do |options|
...
-# Downloads localized metadata for App Store Connect from GlotPress.
+# Downloads localized metadata (intended for App Store Connect) from GlotPress for Jetpack.
 # -----------------------------------------------------------------------------------
 # Usage:
-# bundle exec fastlane download_localized_metadata
+# bundle exec fastlane download_jetpack_localized_app_store_metadata
 #####################################################################################
-desc 'Downloads localized metadata for App Store Connect from GlotPress'
+desc 'Downloads localized Jetpack metadata (for App Store Connect) from GlotPress'
 lane :download_jetpack_localized_app_store_metadata do
```
Comment (AliSoftware) [inline fastlane/Jetpack-Fastfile:167]: While working on this PR, I really felt the pain of those being in different files while they would be related to the same type of action.

While I like the idea of splitting our otherwise-super-large `Fastfile`, I feel that the way we did the split (WP vs JP) was not the best of all, compared to theme-oriented split depending on what category of work the actions are for (release management top-level lanes, vs localization, vs ASC, etc).

See PR description for some ideas about that, though that's definitively for a separate PR. But a separate PR that I'm more and more tempted to address sooner rather than later, especially if I'm gonna continue having to work on the `Fastfiles` while continuing the work on that L10n tooling migration and other related improvements.

File: fastlane/Jetpack-Fastfile [hunk 6]
```diff
@@ -156,41 +156,33 @@ lane :update_jetpack_appstore_strings do |options|
...
-    sh "git add #{jetpack_metadata_glob}"
-    git_commit(
-      path: jetpack_metadata_glob,
-      message: 'Update Jetpack metadata translations'
-    )
-  end
+  git_commit(
+    path: Dir.glob('**/*.txt', base: metadata_directory),
+    message: 'Update Jetpack metadata translations',
+    allow_nothing_to_commit: true
```
Comment (AliSoftware) [inline fastlane/Jetpack-Fastfile:184]: The whole manual logic around `no_files_changed` seemed unnecessary since `git_commit` provides this for free via `allow_nothing_to_commit: true`, which should allow to simplify the whole portion of code quite nicely like this.

**Review by AliSoftware — COMMENTED** [review]
@mokagio while this is still draft, feel free to give a 30% pre-review (sounds familiar? 😝 ) to not only confirm the direction, but also ponder on the ideas I brain-dump'd in the PR description and in this self-review about the path forward

**Comment by mokagio** [comment]
> We should make the `fastlane/{metadata,jetpack_metadata}/en-US/release-notes.txt` be the source of truth where release notes are extracted

Yes! Let's do this. 👍

File: fastlane/Fastfile [hunk 7]
```diff
@@ -458,15 +469,61 @@ platform :ios do
...
   desc 'Downloads localized metadata for App Store Connect from GlotPress'
   lane :download_localized_strings_and_metadata do |options|
-    # WordPress: leverage the default release toolkit setup.
-    # It's not clear in the action name, but this downloads both `.strings`
-    # translations for the app and App Store Connect metadata.
-    ios_update_metadata(options)
-    # Jetpack: use the custom setup. Notice that we don't need extra steps to
-    # download `.strings` because Jetpack uses the same codebase as WordPress.
+    # Download `Localizable.strings` translations used within the app
+    parent_dir_for_lprojs = File.join(PROJECT_ROOT_FOLDER, 'WordPress', 'Resources')
```
Comment (mokagio) [inline fastlane/Fastfile:473]: This is path is also used in top build [`en_lproj` above](https://github.com/wordpress-mobile/WordPress-iOS/pull/18061/files#diff-dff4b99d4e651ab9d7597876ec8dbe92aa934e42c0fb55f4aadd6c106fc96688R440). We could extract it in a constant.

File: fastlane/Fastfile (hunk 3 above)
Comment (mokagio) [reply fastlane/Fastfile:489]: I like the idea of having `default` being also the source of truth for GlotPress. 

Even if we'll never end up localizing hardcoded URLs (I'm sure we have other ways to show users in `country` pages with `language_for_country`, if available, on the web) having them in GlotPress doesn't seem like that much waste.

Relying on a block passed to the localization actions to run experiments seems like a more flexible option than moving files in and out of a folder.

Comment (mokagio) [reply fastlane/Fastfile:489]: > I like the idea of having `default` being also the source of truth for GlotPress.

At that point, we could handle the `en-US` folder by leaving it empty and adding a `.gitkeep` file, so we:

- Wouldn't need to copy files or use `en-GB` as the source for `en-US` (which, by the way, could lead to inappropriate copy, e.g. "colour" vs "color"); and
- Would let `deliver` fallback to the `default` content, making the things full circle: The copy used in ASC for `en-US` is the one from which all (excluding experiments) the other languages have been derived.

File: fastlane/Fastfile (hunk 4 above)
Comment (mokagio) [reply fastlane/Fastfile:510]: > keep that for a later step

Seems wise. I'm not too "worried" about the scenarios as much as I am for MC. But even there, if the file is not found we'll just end up with empty release notes which we can fill manually. That's to say: No need to sweat the parallel PRs coordination too much.

File: fastlane/Jetpack-Fastfile (hunk 6 above)
Comment (mokagio) [reply fastlane/Jetpack-Fastfile:184]: > _(The output above reminds me I should use `git_commit` with `allow_nothing_to_commit: true` like [here](https://github.com/wordpress-mobile/WordPress-Android/pull/15966))_
>
> –https://github.com/wordpress-mobile/WordPress-iOS/pull/17984

👍

**Comment by mokagio** [comment]
_(De-threading from the description and [this comment](https://github.com/wordpress-mobile/WordPress-iOS/pull/18061#discussion_r817080826))_

> Refactor the Fastfile + Jetpack-Fastfile to split the repartition of the lanes differently

I like the setup with the two product-based `Fastfile`s, mainly because it's a nudge to extract Jetpack in its own repo.

But I feel the pain of navigating the large `Fastfile`, even though that doesn't happen so often—and hopefully will happen even left often once we're down with this round of changes.

Also, what changes together should stay together, or, as Uncle Bob [says](https://www.oreilly.com/library/view/97-things-every/9780596809515/ch76.html) "_Gather together those things that change for the same reason, and separate those things that change for different reasons._" Even excluding the changes from this PR, it's reasonable to assume that most `Fastlane` related changes would have to occur across both products. So it makes sense to group the lanes together by what they do rather than who they do it for.

Two things, though:

- Can we have the imported files as plain `.rb` files, so that I don't need to teach my editor each of them should be treated as a Ruby file and not a text file? E.g. `fastlane/Fastfile` (the main one), `fastlane/components/localization.rb` `fastlane/components/app_store_connect.rb`, etc.
- It might be easier to extract "just in time" rather than "just in case." Refactor what you need to work in a way that's simpler for you, but not more. That would leave the repo in a half-finished state, but would also save time. The rest of the refactor could be either a HACK Week project or done whenever the need to work with some of the other lanes arises.

**Comment by AliSoftware** [comment]
💯  on naming the file with à `.rb` extension. I was thinking about `fastlane/lanes/<topic>.rb` rather than `components` but that's just bike-shedding.

As for just-in-time, I'm hesitant, because I'd want this kind of large refactoring to be easily reviewable when we'll do those extraction PRs. Having a PR which _both_ moves an existing lane from one file into another _and_ changes the lane content/code at the same time in the same PR will make it more difficult to review than having a PR focused on moving the code without changing it.

**Comment by mokagio** [comment]
> I was thinking about `fastlane/lanes/<topic>.rb`

That seems good to me, too 👍 

> Having a PR which both moves an existing lane from one file into another and changes the lane content/code at the same time in the same PR will make it more difficult to review than having a PR focused on moving the code without changing it.

Oh, yes, that would not be ideal to review. I should have stated better that my suggestion for just-in-time was entirely about which areas to refactor. But nothing stops us from having [SB diffs](https://medium.com/@kentbeck_7670/bs-changes-e574bc396aaa) limited to a portion of the lanes.

To be clear, I can see the merits in refactoring the _entire_ `Fastfile` structure. I wouldn't push back on that. My only observation is that only refactoring the portion that would make your work easier might save some time.

⚖️ _tradeoffs..._

**Comment by mokagio** [comment]
I bumped this to the next release because we'll be code freezing 19.4 today and this is still a draft.

File: WordPress/Classes/Utility/InfoPListTranslator.m [hunk 8]
```diff
@@ -1,16 +0,0 @@
...
-    NSLocalizedString(@"To take photos or videos to use in your posts.", @"NSCameraUsageDescription: Sentence to justify why the app is asking permission from the user to use is camera.");
-    NSLocalizedString(@"To add photos or videos to your posts.", @"NSPhotoLibraryUsageDescription: Sentence to justify why the app asks permission from the user to access is Media Library.");
-    NSLocalizedString(@"To add photos or videos to your posts.", @"NSPhotoLibraryAddUsageDescription: Sentence to justify why the app asks permission from the user to access is Media Library.");
-    NSLocalizedString(@"Enable microphone access to record sound in your videos.", @"NSMicrophoneUsageDescription: Sentence to justify why the app asks permission from the user to access the device microphone.");
-    NSLocalizedString(@"Save as Draft", @"WordPressDraftActionExtension.CFBundleDisplayName and WordPressDraftActionExtension.CFBundleName: The localised name of the Draft Extension. It will be displayed in the system Share Sheets.");
```
Comment (AliSoftware) [inline WordPress/Classes/Utility/InfoPListTranslator.m:13]: ℹ️  That file was previous used in order for `genstrings` to include those copies — used by our `InfoPlist.strings` — in the translation pipeline. But that wordaround was only partially implemented in practice, because while it indeed allowed those strings to be extracted and included in the `Localizable.strings` when running `genstrings`, the translations that we later pulled from GlotPress during export were never re-injected back into the respective `*.lproj/InfoPlist.strings` files by any automated tooling.

_I think at some point someone copied some of those translations from the `*.lproj/Localizable.strings` to their corresponding `*.lproj/InfoPlist.strings` files **manually** once, and never touched them again, which is why we currently have (potentially / very likely outdated since) translations for our strings in our current `InfoPlist.strings` for various locales…_

---

Now that the new strategy is to merge the manually-maintained `WordPress/Resources/en.lproj/InfoPlist.strings` file within the `Localizable.strings` file that gets imported in GlotPress, then automatically extract back the translations on export, this `InfoPlistTranslator.m` file and the half-workaround it was using is no longer relevant.

The keys that were previously extracted by `genstrings` from that file (with those current English-copy-like keys) will no longer be part of the `Localizable.strings` file, but instead, I've [manually re-inserted the translator hints/comments in the manually-maintained `WordPress/Resources/en.lproj/InfoPlist.strings` (maint target) file](https://github.com/wordpress-mobile/WordPress-iOS/pull/18061/files#diff-3d1519c4554ce94a894e002a96adb518196d2286b4491587437fd48e57ef7c0a) and [`WordPress/WordPressDraftActionExtension/en.lproj/InfoPlist.strings` (action extension target) file](https://github.com/wordpress-mobile/WordPress-iOS/pull/18061/files#diff-9d4c10da21d818a88dfdbc3a7d272e956ea024e6c0a9d09b85ea5a33e49cea05), which will be the new source of truth for those keys from now on.

File: fastlane/Pluginfile [hunk 9]
```diff
@@ -7,8 +7,8 @@ group :screenshots, optional: true do
 end
 
 # This comment avoids typing to switch to a development version for testing.
-# gem 'fastlane-plugin-wpmreleasetoolkit', git: 'git@github.com:wordpress-mobile/release-toolkit', branch: ''
-gem 'fastlane-plugin-wpmreleasetoolkit', '~> 3.0'
+gem 'fastlane-plugin-wpmreleasetoolkit', git: 'git@github.com:wordpress-mobile/release-toolkit', branch: 'l10n/prefix-merge-split-strings'
+# gem 'fastlane-plugin-wpmreleasetoolkit', '~> 3.0'
```
Comment (AliSoftware) [inline fastlane/Pluginfile]: ⚠️ 🚨 To switch back to tag `4.0` once https://github.com/wordpress-mobile/WordPress-iOS/pull/18061 lands and a new release is made

File: fastlane/Fastfile [hunk 10]
```diff
@@ -41,6 +41,8 @@ EXTENSIONS_BUNDLE_IDENTIFIERS = %w[
 
 ALL_BUNDLE_IDENTIFIERS = [APP_STORE_VERSION_BUNDLE_IDENTIFIER, *EXTENSIONS_BUNDLE_IDENTIFIERS]
 
+require_relative 'lanes/localization-lanes'
```
Comment (AliSoftware) [inline fastlane/Fastfile:44]: ℹ️  I plan to move more stuff in that file in a future PR (see idea (2) in the "Next Steps" section of the PR description), especially move all the L10n-related lanes (both from WP and JP) there. But I figured I'd keep that move in a separate PR to keep changes isolated.

File: fastlane/Fastfile (hunk 7 above)
Comment (AliSoftware) [reply fastlane/Fastfile:473]: Good idea 👍  I'll keep that improvement in mind for when I will move all the L10n-related lanes into `lanes/localization-lanes.rb` in a separate, upcoming PR after this one though, where I plan to DRY even more constants and shared paths like that too.

**Comment by AliSoftware** [comment]
@mokagio This PR is now fixed and finalized, and **ready to be reviewed**.

I've marked the PR as "Not ready to merge" because we still need to land https://github.com/wordpress-mobile/release-toolkit/pull/345 and do a new release of the `release-toolkit` first, to be able to [update the `Pluginfile` accordingly before merging](https://github.com/wordpress-mobile/WordPress-iOS/pull/18061#discussion_r826411514).

---

PS: Note that I plan to do subsequent PRs to address the extraction of `Fastfile` lanes to `lanes/*.rb` and some related constant DRY-ing, as well as some other stuff we discussed, separately. I've updated the PR description accordingly to give all the details.

File: fastlane/Fastfile (hunk 7 above)
Comment (mokagio) [reply fastlane/Fastfile:473]: Sounds good to me 👍

File: fastlane/lanes/localization-lanes.rb [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+  # List of `.strings` files manually maintained by developers (as opposed to being automatically extracted from code and generated)
+  # which we will merge into the main `Localizable.strings` file imported by GlotPress, then extract back once we download the translations.
+  # Each `.strings` file to be merged/extracted is associated with a prefix to add the the keys being used to avoid conflicts and differentiate.
+  # See calls to `ios_merge_strings_files` and `ios_extract_keys_from_strings_files` for usage.
+  #
+  MANUALLY_MAINTAINED_STRINGS_FILES = {
+    File.join('WordPress', 'Resources', 'en.lproj', 'InfoPlist.strings') => 'infoplist.', # For now WordPress and Jetpack share the same InfoPlist.strings
+    File.join('WordPress', 'WordPressDraftActionExtension', 'en.lproj', 'InfoPlist.strings') => 'ios-sharesheet.', # CFBundleDisplayName for the "Save as Draft" share action
+    File.join('WordPress', 'WordPressIntents', 'en.lproj', 'Sites.strings') => 'ios-widget.' # Strings from the `.intentdefinition`, used for configuring the iOS Widget
+  }.freeze
```
Comment (mokagio) [inline fastlane/lanes/localization-lanes.rb]: Nitpick: This block of code got shifted one indentation level too much.

```suggestion
# List of `.strings` files manually maintained by developers (as opposed to being automatically extracted from code and generated)
# which we will merge into the main `Localizable.strings` file imported by GlotPress, then extract back once we download the translations.
# Each `.strings` file to be merged/extracted is associated with a prefix to add the the keys being used to avoid conflicts and differentiate.
# See calls to `ios_merge_strings_files` and `ios_extract_keys_from_strings_files` for usage.
#
MANUALLY_MAINTAINED_STRINGS_FILES = {
  File.join('WordPress', 'Resources', 'en.lproj', 'InfoPlist.strings') => 'infoplist.', # For now WordPress and Jetpack share the same InfoPlist.strings
  File.join('WordPress', 'WordPressDraftActionExtension', 'en.lproj', 'InfoPlist.strings') => 'ios-sharesheet.', # CFBundleDisplayName for the "Save as Draft" share action
  File.join('WordPress', 'WordPressIntents', 'en.lproj', 'Sites.strings') => 'ios-widget.' # Strings from the `.intentdefinition`, used for configuring the iOS Widget
}.freeze
```

This suggestion could also be ignored, given there's more work planned on this file.

File: fastlane/lanes/localization-lanes.rb [hunk 12]
```diff
@@ -0,0 +1,63 @@
+# frozen_string_literal: true
+
+#################################################
+# Constants
+#################################################
+
```
Comment (mokagio) [inline fastlane/lanes/localization-lanes.rb:6]: Nitpick. There are only constants in the file right now. 😝 

```suggestion
```

I'm only half joking, though, as it's clear this file will contain more that constants in the future. Maybe you used this style because it's in use in the `Fastfile`. I don't get much value out of these header/separator comments style that's use , though, and I'd rather remove it to have less to scroll through. Even without the comment, the syntax makes the fact that we're looking at constants self-evident.

Needless to say, this is my personal opinion and I'm happy to roll with whatever we feel is best as a team. Consistency trumps personal opinions. Still, I thought it useful to share this little detail.

**Review by mokagio — APPROVED** [review]
Tested via #18135. Looking good.

Approving to speed up the integration process once v4.0.0 of the release toolkit ships.

File: fastlane/lanes/localization-lanes.rb (hunk 11 above)
Comment (AliSoftware) [reply fastlane/lanes/localization-lanes.rb]: Nice catch!

(this is because I initially wrote that constant close to the 2 lanes using it (which are in an indented context), before cut/pasting it here to clean things, and forgetting you remove indentation after paste)

File: fastlane/lanes/localization-lanes.rb (hunk 12 above)
Comment (AliSoftware) [reply fastlane/lanes/localization-lanes.rb:6]: Right. This is because I initially thought I'd move more things to that file (especially lanes) already as part of this PR before deciding I'd do that reorg/move in a separate PR later. So the "Constants" header was differentiate that section from the other "Lanes" section… which ended up not getting moved here just yet.

I'll rearrange all that (and consider dropping the section title if it doesn't help) when I'll move more stuff to that file in my next PR.

Note that I generally agree that large verbose headers add more noise than anything especially if there are too many of them. Which is why I was also thinking of making all the doc comments headers that we have above every single lane in our Fastfiles be shorter, and looking more like YARD doc (or maybe focusing on more useful `desc` blocks) and without the lengthy `#####…` runs in them, to make them more readable, more doc-like and less noisy. And thus only keep 2 or 3 `#####…`-like separators **at most** in whole file to separate big sections like constants vs lanes (using them much like `// MARK: - Section divisor` would be used in Xcode… except if you know of a similar magic marker recognized by RubyMine and/or VSCode we could use for those instead)

Anyway, TL;DR this will be revisited in next PR anyway 🙂

File: fastlane/Pluginfile (hunk 9 above)
Comment (AliSoftware) [reply fastlane/Pluginfile]: I obviously meant "when wordpress-mobile/release-toolkit#345 lands"

**Comment by AliSoftware** [comment]
Thanks for the review! Now just waiting for https://github.com/wordpress-mobile/release-toolkit/pull/345 to be reviewed and approved so I can land both and move forward with the next steps 🙂

**Comment by AliSoftware** [comment]
@mokagio I've updated this PR via 151657b to adopt the latest API changes made in https://github.com/wordpress-mobile/release-toolkit/pull/345#issuecomment-1069471409

As you can see [this makes for a way nicer call site for `ios_merge_strings_files` in the `Fastfile` now](https://github.com/wordpress-mobile/WordPress-iOS/pull/18061/commits/151657b761691fc149a768014468b96c01bd6738#diff-dff4b99d4e651ab9d7597876ec8dbe92aa934e42c0fb55f4aadd6c106fc96688L454-R456) 🎉  and [also simplifies the logic about which files to `git_commit` after keys extraction](https://github.com/wordpress-mobile/WordPress-iOS/pull/18061/commits/151657b761691fc149a768014468b96c01bd6738#diff-dff4b99d4e651ab9d7597876ec8dbe92aa934e42c0fb55f4aadd6c106fc96688L492-R486), so quite a nice win 🎉 

I'll let you check the changes made by that last-minute commit.
If you approve of those latest changes and consider that this PR is ready to merge, feel free to merge it yourself, **but not before you update the reference in the `Pluginfile` to point to the toolkit's `trunk` branch** after you merged https://github.com/wordpress-mobile/release-toolkit/pull/345 (then do a `bundle install` to update the `Gemfile.lock` accordingly of course).

**Comment by mokagio** [comment]
> As you can see [this makes for a way nicer call site for ios_merge_strings_files in the Fastfile now](https://github.com/wordpress-mobile/WordPress-iOS/pull/18061/commits/151657b761691fc149a768014468b96c01bd6738#diff-dff4b99d4e651ab9d7597876ec8dbe92aa934e42c0fb55f4aadd6c106fc96688L454-R456) 🎉 and [also simplifies the logic about which files to git_commit after keys extraction](https://github.com/wordpress-mobile/WordPress-iOS/pull/18061/commits/151657b761691fc149a768014468b96c01bd6738#diff-dff4b99d4e651ab9d7597876ec8dbe92aa934e42c0fb55f4aadd6c106fc96688L492-R486), so quite a nice win 🎉

Yep, noticeable improvement 🎉 🎉
